From 9a92548dd3df1437bd8f5f22b24d32327b15615e Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Sat, 6 May 2023 18:17:28 +0200
Subject: [PATCH 51/55] Revert "maple_tree: Fix potential out-of-bounds access
 in mas_wr_end_piv()"

This reverts commit f4a5e5f0971c22ba17f06f7f20399dbfdfdaeb81.

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 lib/maple_tree.c | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/lib/maple_tree.c b/lib/maple_tree.c
index b37065a6f..18815b89d 100644
--- a/lib/maple_tree.c
+++ b/lib/maple_tree.c
@@ -4262,13 +4262,11 @@ static inline bool mas_wr_slot_store(struct ma_wr_state *wr_mas)
 
 static inline void mas_wr_end_piv(struct ma_wr_state *wr_mas)
 {
-	while ((wr_mas->offset_end < wr_mas->node_end) &&
-	       (wr_mas->mas->last > wr_mas->pivots[wr_mas->offset_end]))
-		wr_mas->offset_end++;
+	while ((wr_mas->mas->last > wr_mas->end_piv) &&
+	       (wr_mas->offset_end < wr_mas->node_end))
+		wr_mas->end_piv = wr_mas->pivots[++wr_mas->offset_end];
 
-	if (wr_mas->offset_end < wr_mas->node_end)
-		wr_mas->end_piv = wr_mas->pivots[wr_mas->offset_end];
-	else
+	if (wr_mas->mas->last > wr_mas->end_piv)
 		wr_mas->end_piv = wr_mas->mas->max;
 }
 
@@ -4425,6 +4423,7 @@ static inline void *mas_wr_store_entry(struct ma_wr_state *wr_mas)
 	}
 
 	/* At this point, we are at the leaf node that needs to be altered. */
+	wr_mas->end_piv = wr_mas->r_max;
 	mas_wr_end_piv(wr_mas);
 
 	if (!wr_mas->entry)
-- 
2.40.1.445.gf85cd430b1

